### DISCUSSION TITLE
query().from_statement() + session.expunge_all()

### DISCUSSION THREAD

**Discussion by mikehagerty** [sqlalchemy/sqlalchemy#7137] [Usage Questions]
I inherited the following code which works in v1.3:
```
statement = text("select net, sta, seedchan, location, min(ondate), max(offdate) "
                 "from channel_data where seedchan in ('SNN', 'SNE', 'BNN', 'BNE', "
                 "'ENN', 'ENE', 'HNN', 'HNE', 'EHN', 'EHE', 'BHN', 'BHE', 'HHN', "
                 "'HHE', 'EH1','EH2','BH1','BH2','HH1','HH2') and  "
                 "samprate in (20, 40, 50, 80, 100, 200) and net=:net and sta=:sta "
                 "group by net, sta, seedchan,location")
statement = statement.columns(Channel_data.net,Channel_data.sta,Channel_data.seedchan,
                              Channel_data.location,Channel_data.ondate,Channel_data.offdate)
rows = session.query(Channel_data).from_statement(statement).params(net=network_code,sta=station_code).all()
```

However, when I try to use this with v1.4 and sessionmaker to do the query:
```
with Session.begin() as session:
    rows = session.query(Channel_data).from_statement(statement).params(net=net, sta=sta).all()
    session.expunge_all()
```

I can print out the cols from the returned Channel_data object, but it crashes if I try to print the object itself:
```
for row in rows:
    print(row.seedchan, row.ondate, row.offdate)   // works
    print(type(row))    // prints <class '...schema_aqms_IR.Channel_data'>
    print(row)    // crashes
```
sqlalchemy.orm.exc.DetachedInstanceError: Instance <Channel_data at 0x125488d90> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)

To solve this I tried setting expire_on_commit to False on the sessionmaker but this did not help:
`Session = sessionmaker(bind=engine, expire_on_commit=False)`

I thought session.expunge_all() was necessary to get the local variable (rows) to be accessible outside of the with Session... context manager.  That is the pattern I use when I do normal query.filter_by() operations, etc, but it seems to be different for query.from_statement() -- e.g., I can access rows outside of the sessionmaker context, regardless of whether or not the session.expunge_all() line is there, and in either case, I can access row.field but not row itself.

Oddly, this code works:
```
with Session.begin() as session:
    rows = session.query(Channel_data).from_statement(statement).params(net=net, sta=sta).all()
    for row in rows:
        print(row)

for row in rows:
    print(row)
```

But if I comment out the print within the sessionmaker context, the print outside fails with the same unbound error:
sqlalchemy.orm.exc.DetachedInstanceError: Instance <Channel_data at 0x119080c70> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)

Any help understanding this nuance would be greatly appreciated.
Thanks!

-Mike

Details: Mac OS 10.14.6, sqlalchemy 1.4.23 + psycopyg2 v2.8.5

**Comment by CaselIT** [comment]
Hi,

It seems that the `__repr__` of the class is trying to access unloaded attributes, but without its definition is hard to know for sure.

**Reply by mikehagerty** [reply]
Thanks - you're right.  I was doing something stupid. I'm not used to querying this way so it didn't occur to me that the select statement wasn't returning all of the columns.  Thanks!
